arm/proofs/bignum_mod_m25519_4.ml

```diff
@@ -0,0 +1,123 @@
+ (`!z x n pc.
+      nonoverlapping (word pc,0x58) (z,8 * 4)
+      ==> ensures arm
+           (\s. aligned_bytes_loaded s (word pc) bignum_mod_m25519_4_mc /\
+                read PC s = word pc /\
+                C_ARGUMENTS [z; x] s /\
+                bignum_from_memory (x,4) s = n)
+           (\s. read PC s = word (pc + 0x54) /\
+                bignum_from_memory (z,4) s = n MOD m_25519)
+          (MAYCHANGE [PC; X2; X3; X4; X5; X6; X7; X8; X9] ,,
+           MAYCHANGE SOME_FLAGS ,,
+           MAYCHANGE [memory :> bignum(z,4)])`,
+  MAP_EVERY X_GEN_TAC [`z:int64`; `x:int64`; `m:num`; `pc:num`] THEN
+  REWRITE_TAC[C_ARGUMENTS; C_RETURN; SOME_FLAGS; NONOVERLAPPING_CLAUSES] THEN
+  DISCH_THEN(REPEAT_TCL CONJUNCTS_THEN ASSUME_TAC) THEN
+
+  BIGNUM_TERMRANGE_TAC `4` `m:num` THEN
+  REWRITE_TAC[BIGNUM_FROM_MEMORY_BYTES] THEN ENSURES_INIT_TAC "s0" THEN
+  BIGNUM_DIGITIZE_TAC "m_" `read (memory :> bytes (x,8 * 4)) s0` THEN
+
+  ARM_ACCSTEPS_TAC BIGNUM_MOD_M25519_4_EXEC (1--21) (1--21) THEN
+  RULE_ASSUM_TAC(REWRITE_RULE
+   [REAL_BITVAL_NOT; ADD_CLAUSES; VAL_WORD_BITVAL]) THEN
+
+  SUBGOAL_THEN `carry_s15 <=> m < m_25519` SUBST_ALL_TAC THENL
+   [MATCH_MP_TAC FLAG_FROM_CARRY_LT THEN EXISTS_TAC `256` THEN
+    EXPAND_TAC "m" THEN REWRITE_TAC[m_25519; GSYM REAL_OF_NUM_CLAUSES] THEN
+    REWRITE_TAC[REAL_BITVAL_NOT] THEN
+    ACCUMULATOR_ASSUM_LIST(MP_TAC o end_itlist CONJ o DECARRY_RULE) THEN
+    DISCH_THEN(fun th -> REWRITE_TAC[th]) THEN BOUNDER_TAC[];
+    ALL_TAC] THEN
+
+  ENSURES_FINAL_STATE_TAC THEN ASM_REWRITE_TAC[] THEN
+  CONV_TAC(LAND_CONV BIGNUM_EXPAND_CONV) THEN
+  ASM_REWRITE_TAC[] THEN DISCARD_STATE_TAC "s21" THEN
+
+  W(MP_TAC o PART_MATCH (lhand o rand) MOD_CASES o rand o snd) THEN
+  ANTS_TAC THENL
+   [REWRITE_TAC[m_25519] THEN ASM_ARITH_TAC; DISCH_THEN SUBST1_TAC] THEN
+  COND_CASES_TAC THEN ASM_REWRITE_TAC[] THEN
+
+  RULE_ASSUM_TAC(REWRITE_RULE[NOT_LT]) THEN
+  ASM_SIMP_TAC[GSYM REAL_OF_NUM_EQ; GSYM REAL_OF_NUM_SUB] THEN
+  EXPAND_TAC "m" THEN REWRITE_TAC[GSYM REAL_OF_NUM_ADD] THEN
+  REWRITE_TAC[GSYM REAL_OF_NUM_POW; GSYM REAL_OF_NUM_MUL] THEN
+  MATCH_MP_TAC EQUAL_FROM_CONGRUENT_REAL THEN
+  MAP_EVERY EXISTS_TAC [`256`; `&0:real`] THEN ASM_REWRITE_TAC[] THEN
+  CONJ_TAC THENL [BOUNDER_TAC[]; ALL_TAC] THEN
+  CONJ_TAC THENL
+   [ASM_REWRITE_TAC[REAL_OF_NUM_ADD; REAL_OF_NUM_POW; REAL_OF_NUM_MUL] THEN
+    ASM_REWRITE_TAC[REAL_SUB_LE; REAL_OF_NUM_LE] THEN
+    MATCH_MP_TAC(REAL_ARITH `x:real < y ==> x - &n < y`) THEN
+    REWRITE_TAC[REAL_OF_NUM_LT] THEN ASM_ARITH_TAC;
+    ALL_TAC] THEN
+  CONJ_TAC THENL [REAL_INTEGER_TAC; ALL_TAC] THEN
+
+  ACCUMULATOR_POP_ASSUM_LIST(MP_TAC o end_itlist CONJ o DESUM_RULE) THEN
+  REWRITE_TAC[REAL_BITVAL_NOT; m_25519] THEN
+  DISCH_THEN(fun th -> REWRITE_TAC[th]) THEN
+  CONV_TAC(RAND_CONV REAL_POLY_CONV) THEN REAL_INTEGER_TAC);;
+
+let BIGNUM_MOD_M25519_4_SUBROUTINE_CORRECT = time prove
+ (`!z x n pc returnaddress.
+      nonoverlapping (word pc,0x58) (z,8 * 4)
```

Comment:
Would the specification be more readable if we use `length bignum_mod_m25519_4_mc` instead of constant `0x58`? This could be applied to other files.

Reply:
Yes, this is certainly true; it might well make sense to write all the "code size" specs like this, which would be more readable and less brittle (though a bit more verbose, particularly in code + data cases with data tables included in the size). I will not change it right here, but we should perhaps add this to the to-do list of general improvements to s2n-bignum.

Reply:
Thank you for the careful general check and extracting the specifications! This is probably exactly the right thing for people who are not concerned with the inner workings of the HOL Light proofs to look at in order to decide whether the functionality is right.

Reply:
I've fixed the issue June noted above, as well as switching edwards25519_scalarmul* functions to the divstep-based modular inverse in line with the curve25519 scalar multiplications.